sretoolbox/utils/multiprocess.py

```diff
@@ -0,0 +1,71 @@
+# Copyright 2021 Red Hat
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Multiprocessing abstractions.
+"""
+
+from multiprocessing import Pool
+
+
+def run(func, iterable, process_pool_size,
+        return_exceptions=False, **kwargs):
+    """
+    run_processes executes a function for each item in the input iterable.
+    execution will be done in a processpool according to the input
+    process_pool_size.  kwargs are passed to the input function
+    (optional). If return_exceptions is true, any exceptions that may
+    have happened in each thread are returned in the return value,
+    allowing the caller to get as much work done as possible.
+    """
+    if return_exceptions:
+        tracer = _catching_traceback
+    else:
+        tracer = _full_traceback
+
+    task_list = [(func, [i], kwargs) for i in iterable]
+    with Pool(process_pool_size) as pool:
+        return pool.map(tracer, task_list)
+
+
+def _catching_traceback(spec):
+    try:
+        func, args, kwargs = spec
+        return func(*args, **kwargs)
+    # pylint: disable=broad-except
+    except Exception as details:
+        return details
+    except SystemExit as details:
+        if int(details.code) != 0:
+            return details
+        return None
```

Comment:
This line is the only thing I'm not sure about in this whole module. Say we are using `_catching_traceback` and we have a `SystemExit(1)` in one of the processes, the final results list will look line this:
```python
[1, SystemExit(1), 3, 4, 5]
```

On the other hand, for a `SystemExit(0)`, the final results list looks like this:
```pyhton
[1, None, 3, 4, 5]
```
I'm trying to think of a case where having `None` is more informative and relevant than having the actual exception, like this:
```python
[1, SystemExit(0), 3, 4, 5]
```
Do you have a case in mind? Maybe you're processing the results list somewhere else, checking for the type of each item?

Reply:
the behaviour of the `threading` module is the same so everything we discuss here is relevant for existing code as well.

### speaking in the context of q-r

`SystemExit(0)` is special in a way that it is not necessarily indicating an exception in q-r. `sys.exit(0)` is used in various integrations to end it successfully. not necessarily the nicest way to terminate a process/thread mid-method down in a call stack but in the context of q-r it translates to "end the execution of this integration with no result" - hence `None`.  you made a very valid point that this could be up to the consuming client to interpret. q-r could easily interpret a `SystemExit(0)` as "no error and no result". 

if we would hand back `SystemExit()` exceptions in `return_exceptions=True` cases, we would need to make minor adjustments in q-r because we are checking for the type there. low risk + low effort. i'm absolutely willing to do that.

### speaking generally

as i said, this behaviour is currently also present in the `threaded` module. if we want to consistently return these `SystemExit` exception for the `threaded` and `multiprocessors` module while `return_exceptions=True`, we need to check all our projects that are using this feature. i'm only aware of `q-r`. are there others you are aware @apahim @rporres? 

i'm not too worried about any usecases of `threaded.run(..., return_exceptions=True)` to break on `SystemExit` because it was broken [until recently](https://github.com/app-sre/sretoolbox/pull/87) anyways, rendering the ThreadPool unresponsive in such situations.

so if it would be ok for everyone and if this will not cause issues in other codebases, i would go ahead with @apahim suggestion to plainly return `SystemExit` exceptions, document it and adapt `q-r` to only handle `SytemExit(1)` accordingly.

Reply:
Got it, thank you for looking into that! I'm not aware of any other cases where that would be of any harm.

I'm ok with handling it now, but I'm also fine with getting this merged as is, and either not handling it at all or handling it in a separate PR/Story :)

Reply:
I'm not aware of this being used anywhere else, so it makes sense to me to go with @apahim's suggestion.